### PR TITLE
Add support for noexcept keyword in C++ class methods

### LINKED ISSUE #11871: embind does not support class methods qualified with "noexcept"
This can be fixed by adding two new instantiations of `RegisterClassMethod` to `bind.h`, as seen here:

https://github.com/google/filament/pull/2940/commits/cddc82dceb7d968eec8de5f927027f7cfe0a9d8d

We should probably also update emscripten's unit tests to cover these cases.

To give some context, this will fix the following clang errors, which we ran into when upgrading the Filament build from C++14 to C++17:

```
/Users/prideout/github/emsdk/upstream/emscripten/system/include/emscripten/bind.h:1553:13: error: implicit instantiation of undefined template 'emscripten::internal::RegisterClassMethod<void (filament::View::*)(filament::View::AmbientOcclusion) noexcept>'
            invoker::template invoke<ClassType, Policies...>(methodName, callable);
            ^
../../web/filament-js/jsbindings.cpp:566:6: note: in instantiation of function template specialization 'emscripten::class_<filament::View, emscripten::internal::NoBaseClass>::function<emscripten::internal::DeduceArgumentsTag, void (filament::View::*)(filament::View::AmbientOcclusion) noexcept>' requested here
    .function("setAmbientOcclusion", &View::setAmbientOcclusion)
     ^
/Users/prideout/github/emsdk/upstream/emscripten/system/include/emscripten/bind.h:1285:16: note: template is declared here
        struct RegisterClassMethod;
               ^
```

### PR DESCRIPTION
**Pull request by harichards**
Closes #11871.

This PR is based off of the work done in #10613, except it also adds support for `noexcept` when using `class_::property()`. The conversation last spoke about merging the template specializations to reduce duplicate code. This is possible using deduction of the `noexcept` keyword like so:

```cpp
template<typename ClassType, typename ReturnType, typename... Args, bool IsNoexcept>
struct RegisterClassMethod<ReturnType (ClassType::*)(Args...) noexcept(IsNoexcept)> {
    template <typename CT, typename... Policies>
    static void invoke(const char* methodName, ReturnType (ClassType::*memberFunction)(Args...) noexcept(IsNoexcept)) {
        ...
    }
};
```

However, `noexcept` deduction is a non-standard extension of GCC and Clang introduced with C++17. I'm not sure this can be done without sacrificing support for some compilers and versions of C++ (see https://devblogs.microsoft.com/oldnewthing/20200714-00/?p=103981).

### CODE REVIEW

**Comment by harichards** [comment]
@kripken

Checking in again, hoping someone has time to review this change. Thanks!

**Comment by sbc100** [comment]
@RReverser

File: system/include/emscripten/bind.h [hunk 1]
```diff
@@ -634,6 +634,12 @@ struct GetterPolicy<GetterReturnType (GetterThisType::*)() const> {
     }
 };
 
+#ifdef __cpp_noexcept_function_type
```
[the remark this answers is not in the document]
Comment (harichards) [reply system/include/emscripten/bind.h:637]: Yes, you can see the macro mentioned here: https://en.cppreference.com/w/cpp/feature_test

As of the most recent commit, this change **should not** break compatibility with any version of C++. I have tested this by successfully compiling and running code using this change for both C++11 and C++17. I can add this as a unit test.

**Comment by david-fong** [comment]
fixes #9485

**Comment by david-fong** [comment]
Let me know if I can help with this. Looking at [main/tests/embind/embind_test.cpp](https://github.com/emscripten-core/emscripten/blob/main/tests/embind/embind_test.cpp), would it be good to add something like what's there for "SmallClass", but instead "NoexceptClass" and give it a noexcept member function and also a noexcept free-function taking taking NoexceptClass as its first parameter?

**Comment by brendandahl** [comment]
Included in #17140.

**Pull request closed without merging**